pkg/server/requestLimiter.go

```diff
@@ -0,0 +1,53 @@
+// Copyright Contributors to the Open Cluster Management project
+
+package server
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	klog "k8s.io/klog/v2"
+
+	"github.com/gorilla/mux"
+	"github.com/stolostron/search-indexer/pkg/config"
+)
+
+var pendingRequests = map[string]time.Time{}
+var pendingLock = sync.RWMutex{}
+
+// Checks if we are able to accept the incoming request.
+func requestLimiterMiddleware(next http.Handler) http.Handler {
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		params := mux.Vars(r)
+		clusterName := params["id"]
+
+		klog.V(6).Info("Checking if we can process incoming request. Existing pending requests: ", len(pendingRequests))
+
+		if t, exists := pendingRequests[clusterName]; exists {
+			klog.Warningf("Rejecting request from %s because there's a previous request processing. Duration: %s",
+				clusterName, time.Since(t))
+			http.Error(w, "A previous request from this cluster is processing, retry later.", http.StatusTooManyRequests)
+			return
+		}
+
+		if len(pendingRequests) >= config.Cfg.RequestLimit && clusterName != "local-cluster" {
+			klog.Warningf("Too many pending requests (%d). Rejecting sync from %s", len(pendingRequests), clusterName)
+			http.Error(w, "Indexer has too many pending requests, retry later.", http.StatusTooManyRequests)
```

Comment:
Should we add the cluster name to pending requests before returning?

Reply:
Or do we not keep track of rejected requests?

Reply:
Ok, pending here means currently processing, I guess?

Reply:
This line is the response message sent to the cluster, I don't think the name is needed here.
The line above is the log and it has the cluster name.

Reply:
Oh, did you mean the `pendingRequest` map?
No, because we won't be processing the request.

Reply:
Would it help if I rename it to `outstandingRequests` or `requestsProcessing` ?

Reply:
yes, requestsProcessing would be good.

Reply:
changed